client/vscode/src/settings/invalidation.ts

```
@@ -0,0 +1,51 @@
+import * as vscode from 'vscode'
+
+import { invalidateClient } from '../backend/requestGraphQl'
+import { VSCEStateMachine } from '../state'
+
+/**
+ * Listens for Sourcegraph URL or access token changes and invalidates the GraphQL client
+ * to prevent data "contamination" (e.g. sending private repo names to Cloud instance).
+ */
+export function invalidateContextOnSettingsChange({
+    context,
+    stateMachine,
+}: {
+    context: vscode.ExtensionContext
+    stateMachine: VSCEStateMachine
+}): void {
+    function disposeAllResources(): void {
+        for (const subscription of context.subscriptions) {
+            subscription.dispose()
+        }
+    }
+
+    context.subscriptions.push(
+        vscode.workspace.onDidChangeConfiguration(config => {
+            if (config.affectsConfiguration('sourcegraph.accessToken')) {
+                invalidateClient()
+                disposeAllResources()
+                stateMachine.emit({ type: 'access_token_change' })
+                vscode.window
+                    .showInformationMessage(
+                        'Restart VS Code to use the Sourcegraph extension after access token change.'
+                    )
+                    .then(
+                        () => {},
+                        () => {}
+                    )
```

Why do we keep adding these empty promise listeners everywhere? It seems like it would make it very easy to swallow any issues. Should we instead convert to an async function and call "await"?

---

The intention is to swallow errors, or to basically treat `vscode.window.showInformationMessage` as fire-and-forget. Any idea on how to make that more explicit? e.g. just leave it unhandled and disable the unhandled promise lint rule w/ a comment?

---

> The intention is to swallow errors, or to basically treat vscode.window.showInformationMessage as fire-and-forget. Any idea on how to make that more explicit? e.g. just leave it unhandled and disable the unhandled promise lint rule w/ a comment?

I suppose the lint rule is there for a reason. If we don't catch errors it could very easily break in the future (e.g. someone inserts a syntax error or so).

If you're absolutely sure this can break and we don't care, it's alright to leave as-is but my comment was more about this pattern in general as I've seen it a couple of times in this PR.

Feel free to ignore this feedback btw, just wanted to raise my concern mostly. 🙂

---

This is the most efficient way I know of for swallowing errors in promises, so it's great at that.
But I agree with Philip that without any comment, the next developer might raise an eyebrow and add some logging in the catch block.
So I'd probably mark it with a comment along the lines of  `// Swallowing all errors: this is always safe because ...`

---

Thank you both! I should document my reasoning (and feel free to challenge it here): `vscode.window.showInformationMessage` only has a meaningful return value when you call it with a list of options, which is why callback #1 is a noop. If it fails despite not even prompting the user for input, I assume that something about the VS Code extension host is borked, and retrying would be futile; why else should an API akin to `window.alert` fail?